### PR TITLE
Manage reduced service

### PR DESCRIPTION
**Pull request by azime**
Manage REDUCED_SERVICE effect:

Example:
image

Jira: [NAV-893](https://navitia.atlassian.net/browse/NAV-893)

### CODE REVIEW

File: source/kraken/apply_disruption.cpp [hunk 1]
```diff
@@ -387,34 +402,45 @@ struct add_impacts_visitor : public apply_impacts_visitor {
                 LOG4CPLUS_TRACE(log, "impacted vj : " << vj_uri << " not found in data. I ignore it.");
                 continue;
             }
+            if (impacted_vj.impacted_ranks.empty()) {
+                LOG4CPLUS_TRACE(log, "impacted vj : " << vj_uri << " without impacted_ranks data. I ignore it.");
+                continue;
+            }
+            const auto begin = impacted_vj.impacted_ranks.begin();
+            const auto end = std::next(begin, impacted_vj.impacted_ranks.size() - 1);
```
Comment (AzHicham) [inline source/kraken/apply_disruption.cpp:410]: May be use  `const auto end = impacted_vj.impacted_ranks.rbegin() `

**Review by AzHicham — APPROVED** [review]

**Review by woshilapin — APPROVED** [review]
Didn't see anything obviously weird, but I'm really not confident of my understanding. Another pair of eyes is probably good :eyes:

**Comment by pbench** [comment]
Could you re-open the PR from a branch of https://github.com/hove-io/navitia/ instead of your fork https://github.com/azime/navitia/ ?
I think it will turn the CI to :green_circle:

File: source/kraken/apply_disruption.cpp [hunk 2]
```diff
@@ -387,34 +402,45 @@ struct add_impacts_visitor : public apply_impacts_visitor {
...
                 // We need to get the associated base stop_time to compare its rank
                 const auto base_st = st.get_base_stop_time();
-                // stop is ignored if its stop_point is not in impacted_stops
-                // if we don't find an associated base we keep it
-                if (base_st && impacted_vj.impacted_ranks.count(base_st->order())) {
-                    LOG4CPLUS_TRACE(log, "Ignoring stop " << st.stop_point->uri << "on " << vj->uri);
-                    if (first_sa == true) {
-                        break;
+                if (impact->severity->effect == nt::disruption::Effect::REDUCED_SERVICE
+                    && impacted_vj.impacted_ranks.count(base_st->order())) {
```
Comment (pbench) [inline source/kraken/apply_disruption.cpp:420]: the previous version of the code tested if `base_st != null`. You don't need this check anymore ?

File: source/kraken/apply_disruption.cpp [hunk 3]
```diff
@@ -387,34 +402,45 @@ struct add_impacts_visitor : public apply_impacts_visitor {
...
             for (const auto& st : vj->stop_time_list) {
                 // We need to get the associated base stop_time to compare its rank
                 const auto base_st = st.get_base_stop_time();
-                // stop is ignored if its stop_point is not in impacted_stops
-                // if we don't find an associated base we keep it
-                if (base_st && impacted_vj.impacted_ranks.count(base_st->order())) {
-                    LOG4CPLUS_TRACE(log, "Ignoring stop " << st.stop_point->uri << "on " << vj->uri);
-                    if (first_sa == true) {
-                        break;
+                if (impact->severity->effect == nt::disruption::Effect::REDUCED_SERVICE
```
Comment (pbench) [inline source/kraken/apply_disruption.cpp:419]: If I understand correctly, you are now doing two different thing in this loop : 

- if effect == No_Service : you want to keep all the stop_times until you find one contained in `impacted_ranks`
- if effect == Reduced_Service : you want to keep all the stop_times, but modify the ones that are contained in `impacted_ranks`, except `rs.start_stop `and `rs.end_stop`

To make things easier to read, I suggest putting the `if` before the loop, and have two specialized loops if the if/else branches

**Comment by pbench** [comment]
reopened in https://github.com/hove-io/navitia/pull/3756

**Pull request closed without merging**